src/state/slices/portfolioSlice/portfolioSlice.ts

```diff
@@ -53,38 +53,42 @@ export const portfolio = createSlice({
         getMixPanel()?.track(MixPanelEvents.DisconnectWallet)
       }
     },
-    upsertAccountMetadata: (state, { payload }: { payload: AccountMetadataById }) => {
-      moduleLogger.debug('upserting account metadata')
-      state.accountMetadata.byId = {
-        ...state.accountMetadata.byId,
-        ...payload,
-      }
+    upsertAccountMetadata: {
+      reducer: (state, { payload }: { payload: AccountMetadataById }) => {
+        state.accountMetadata.byId = merge(state.accountMetadata.byId, payload)
+        state.accountMetadata.ids = Object.keys(state.accountMetadata.byId)
 
-      state.accountMetadata.ids = Object.keys(state.accountMetadata.byId)
+        if (!state.walletId) return // realistically, at this point, we should have a walletId set
+        const existingWalletAccountIds = state.wallet.byId[state.walletId] ?? []
+        const newWalletAccountIds = Object.keys(payload)
+        // keep an index of what account ids belong to this wallet
+        state.wallet.byId[state.walletId] = Array.from(
+          new Set(existingWalletAccountIds.concat(newWalletAccountIds)),
+        )
+      },
 
-      if (!state.walletId) return // realistically, at this point, we should have a walletId set
-      const existingWalletAccountIds = state.wallet.byId[state.walletId] ?? []
-      const newWalletAccountIds = Object.keys(payload)
-      // keep an index of what account ids belong to this wallet
-      state.wallet.byId[state.walletId] = Array.from(
-        new Set([...existingWalletAccountIds, ...newWalletAccountIds]),
-      )
+      // Use the `prepareAutoBatched` utility to automatically
+      // add the `action.meta[SHOULD_AUTOBATCH]` field the enhancer needs
+      prepare: prepareAutoBatched<AccountMetadataById>(),
     },
-    upsertPortfolio: (state, { payload }: { payload: Portfolio }) => {
-      moduleLogger.debug('upserting portfolio')
-      // upsert all
-      state.accounts.byId = { ...state.accounts.byId, ...payload.accounts.byId }
-      const accountIds = Array.from(new Set([...state.accounts.ids, ...payload.accounts.ids]))
-      state.accounts.ids = accountIds
+    upsertPortfolio: {
+      reducer: (state, { payload }: { payload: Portfolio }) => {
+        moduleLogger.debug('upserting portfolio')
+        // upsert all
+        state.accounts.byId = merge(state.accounts.byId, payload.accounts.byId)
+        const accountIds = Array.from(new Set(state.accounts.ids.concat(payload.accounts.ids)))
+        state.accounts.ids = accountIds
```

Comment:
could we get away with the same approach you've used above to avoid concat and uniq?

```suggestion
        state.accounts.ids = Object.keys(state.accounts.byId)
```